### ISSUE TITLE
Post timeout errors

### ISSUE THREAD

**Issue by Sophietje** [TNO/knowledge-engine#215]
In GitLab by @michel.hours.inetum.world on May 21, 2021, 11:28

Today I'm experiencing a strange behavior when I'm trying to post some data. The data is well received on the react site but on the post side there is a timeout error. The same issue occurs using either a Java program or Postman.

**Comment by Sophietje** [comment]
In GitLab by @barry.nouwt.tno.nl on May 21, 2021, 11:45

Hi @michel.hours.inetum.world , thanks for letting us know.

We found a problem where all the threads of the webserver were getting blocked. We are working on a new release that should solve these issues, until then I will restart the server. 

I assume the problem will for now be fixed.

**Comment by Sophietje** [comment]
In GitLab by @barry.nouwt.tno.nl on May 21, 2021, 11:53

@michel.hours.inetum.world We also noticed multiple knowledge bases that register graph patterns like "?s ?p ?o" (i.e. from the examples we provided). If this is the case, it might occur that you ask a question or post some data and there are multiple KBs available that can answer or want to react to that type of data (i.e. they use a matching graph pattern). This means that you may not receive a request for data on your KB until one of the others has answered or reacted, or you might get a request, but you do not see the expected reaction in your other KB, because the Interoperability layer is waiting for the other matching KBs to answer/react.

We have an issue #95 which would allow you to instruct the Knowledge Engine to not wait indefinitely for an answer, but this is still on our todo list. Until then, we recommend using more specific graph patterns for testing. For example:

```
{
  "knowledgeInteractionType": "ReactKnowledgeInteraction",
  "argumentGraphPattern": "?s <http://inetum.world/hasValue> ?o ."
}
```

**Comment by Sophietje** [comment]
In GitLab by @johan.kumps.realdolmen.com on May 25, 2021, 19:50

Could you restart the server again? I also experience this problem starting this afternoon.

Thanks,
Johan,

**Comment by Sophietje** [comment]
In GitLab by @barry.nouwt.tno.nl on May 27, 2021, 06:49

Hi @johan.kumps.realdolmen.com, can you give us more information on when the timeout happened? 

Currently, we have no indications that the Knowledge Engine is instable. All our test examples are still working.

When did the timeout occur? Was it a timeout from your REST client, or from the server-side? And which operation did you call when the timeout occurred?

Regards, Barry

**Comment by Sophietje** [comment]
In GitLab by @johan.kumps.realdolmen.com on May 27, 2021, 06:59

I just tried the eu.interconnectproject.knowledge_engine.rest.api.client_example.post_react.PostingClient and ReactingClient example from within the KE project and the timeout occurs. Monday morning this was working fine. In the afternoon the timeout started to appear.

I'm using the Rest client indeed. 

`var client = new RestApiClient("https://ke.interconnectproject.eu/rest");`

**Comment by Sophietje** [comment]
In GitLab by @barry.nouwt.tno.nl on May 27, 2021, 09:45

Hey @johan.kumps.realdolmen.com, I investigated it and notice the timeout as well when I run the Posting and Reacting clients. The problem only occurs when running on the cloud KE at ke.interconnectproject.eu/rest...the local version is working fine. The FAQ below explains why this happens when you use very general graph patterns like `?s ?p ?o` like we do in our test clients.

*Question*: I use a very generic graph pattern like `?s ?p ?o` for my PostKnowledgeInteraction, but my other KB does not get a request. Or, you do get a request, but your post is not returning with the results.
* *Answer*: We noticed multiple knowledge bases that register graph patterns like "?s ?p ?o" (i.e. from the examples we provided). If this is the case, it might occur that you ask a question or post some data and there are multiple KBs available that can answer or want to react to that type of data (i.e. they use a matching graph pattern). This means that you may not receive a request for data on your KB until one of the others has answered or reacted, or you might get a request, but you do not see the expected reaction in your other KB, because the Interoperability layer is waiting for the other matching KBs to answer/react. We have an issue #95 which would allow you to instruct the Knowledge Engine to not wait indefinitely for an answer, but this is still on our todo list. Until then, we recommend using more specific graph patterns for testing. For example: `{ "knowledgeInteractionType": "ReactKnowledgeInteraction", "argumentGraphPattern": "?s http://inetum.world/hasValue ?o ." }`

Try using a more specific argument and result graph patterns and also keep in mind that this code assumes their two Smart Connectors do not yet exist (but they do not remove the SCs themselves). I've tested it and it indeed solved the timeout if I make sure the SCs do not yet exist in the cloud version and I use less generic graph patterns. I've created issue #206 to improve these clients.

Hope this helps and, again, thanks for the notification!

**Comment by Sophietje** [comment]
In GitLab by @johan.kumps.realdolmen.com on May 27, 2021, 09:50

Hi @barry.nouwt.tno.nl , I understand that these very generic graph patterns are not that the best ones but as mentioned in [](https://gitlab.inesctec.pt/interconnect/knowledge-engine/-/issues/204) I get the same errors with much more complex patterns. Also only when using the cloud KE. A local KE is fine. When I use a differen KB_ID when starting the client :

`private static final String KB_ID = "https://www.interconnectproject.eu/knowledge-engine/knowledgebase/example/a-measurements-posting-kb" + "-" + UUID.randomUUID();
`

The timeout indeed is gone. So in this case the cloud KE is reacting correctly.

Does this mean that when using the KE in the Cloud we should not create the smart connector nor register the knowledge interactions each time. This would be fine as we could doe that at deploy time of our application.

Johan,

**Comment by Sophietje** [comment]
In GitLab by @barry.nouwt.tno.nl on May 27, 2021, 12:55

Hi @johan.kumps.realdolmen.com, thanks for letting us know this was indeed related to what we expected.

Ideally, I think you application should do the following:
1. start application
2. check if knowledge base already exists in the Knowledge Engine
3. if not, register new Knowledge Base
4. check if necessary knowledge interactions already exist
5. if not, register the necessary Knowledge Interactions
6. ...exchange the data...
7. while stopping application
8. remove your knowledge base from the Knowledge Engine
9. shutdown

As discussed during the WP5 sync meeting, the generic adapter team with @aleksandar.tomcic.vizlore.com is working on doing this automatically for you.

I close this issue now, ok? If you have new remarks/questions, create a new issue.

**Comment by Sophietje** [comment]
In GitLab by @johan.kumps.realdolmen.com on May 27, 2021, 13:16

Thanks @barry.nouwt.tno.nl ! Again the timeout is there when using the KE in the cloud using a different KB_ID in the Post part:

```
2021-05-27 15:20:02:598 +0200 [main] INFO PostingClient - Made new KI with ID https://www.interconnectproject.eu/knowledge-engine/knowledgebase/example/a-measurements-posting-kb-963d1a09-5374-42db-8381-df56ea150ff0/interaction/413ed427-8e56-4b4a-9ff4-1a517a3554df
2021-05-27 15:20:03:603 +0200 [main] INFO PostingClient - Sending POST: [Binding [map={p=<https://www.tno.nl/example/prop1>, d=<https://www.tno.nl/example/device1>, pwr=10, m=<https://www.tno.nl/example/meas1>, ts='2002-0530T09:00:00^^xsd:timestamp'}]]
java.net.SocketTimeoutException: timeout
	at okhttp3.internal.http2.Http2Stream$StreamTimeout.newTimeoutException(Http2Stream.kt:677)
	at okhttp3.internal.http2.Http2Stream$StreamTimeout.exitAndThrowIfTimedOut(Http2Stream.kt:686)
	at okhttp3.internal.http2.Http2Stream.takeHeaders(Http2Stream.kt:143)
	at okhttp3.internal.http2.Http2ExchangeCodec.readResponseHeaders(Http2ExchangeCodec.kt:96)
	at okhttp3.internal.connection.Exchange.readResponseHeaders(Exchange.kt:106)
	at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.kt:79)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.kt:34)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.kt:95)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.kt:83)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.kt:76)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:201)
	at okhttp3.internal.connection.RealCall.execute(RealCall.kt:154)
	at eu.interconnectproject.knowledge_engine.rest.api.client_example.RestApiClient.postPost(RestApiClient.java:267)
	at eu.interconnectproject.knowledge_engine.rest.api.client_example.measurements_post_react.PostingClient.main(PostingClient.java:87)
Exception in thread "main" java.lang.RuntimeException: Could not POST the POST message.
	at eu.interconnectproject.knowledge_engine.rest.api.client_example.RestApiClient.postPost(RestApiClient.java:276)
	at eu.interconnectproject.knowledge_engine.rest.api.client_example.measurements_post_react.PostingClient.main(PostingClient.java:87)

Process finished with exit code 1
```

This behavior is also there when running the KE locally. Sequence :

1. start KE
2. start ReactingClient creating the smart connector and registering KIs with UUID in the KB_ID
3. Start PostingClient --> request received in ReactingClient without errors
4. Restart ReactingClient creating the smart connector and registering KIs with UUID in the KB_ID
5. Restart PostingClient --> request received in ReactingClient but error in PostingClient

I did not restart the KE. If do this, all works fine. This is strange as adding a completely separate SC and KI I do not expect this.

**Comment by Sophietje** [comment]
In GitLab by @barry.nouwt.tno.nl on May 27, 2021, 13:26

Strange...

**Comment by Sophietje** [comment]
In GitLab by @barry.nouwt.tno.nl on May 27, 2021, 13:49

Thanks for the sequence. I'll test it locally as well and see what happens.

**Comment by Sophietje** [comment]
In GitLab by @barry.nouwt.tno.nl on May 27, 2021, 14:02

I can reproduce it locally with the clients you provided (after adding the UUID). I will check what is going wrong.

**Comment by Sophietje** [comment]
In GitLab by @barry.nouwt.tno.nl on May 27, 2021, 14:14

@johan.kumps.realdolmen.com Do you stop/terminate the reacting client from step 2 before starting a new one?

1. start KE
1. start ReactingClient creating the smart connector and registering KIs with UUID in the KB_ID
1. Start PostingClient --> request received in ReactingClient without errors
1. *Stop ReactingClient from step 2.*
1. Restart ReactingClient creating the smart connector and registering KIs with UUID in the KB_ID
1. Restart PostingClient --> request received in ReactingClient but error in PostingClient

If I do step 4 in the above sequence, the problem occurs. If I skip step 4, it will work without problems. In the logs of the posting client, you can also see why this is the case. The Knowledge Engine has two reacting clients registered that have indicated they want to react to what the postingclient posts. The first time, there is only a single reacting client. So, the second time it posts, the KE will contact the smart connector of _both_ reacting clients, but if one of the reacting *clients* was already stopped, it can no longer react to the handle request from its Smart Connector (which is still available) and the Knowledge Engine will keep waiting until it reacts. This takes too long for the PostingClient and it will throw a TimeoutException.

So, if you keep the first reaction client running and then start the second. Do you not see the timeout?

**Comment by Sophietje** [comment]
In GitLab by @johan.kumps.realdolmen.com on May 27, 2021, 14:29

I do stop the ReactingClient and the PostingClient. That is what I meant by restarting.

>> If I do step 4 in the above sequence, the problem occurs. If I skip step 4, it will work without problems.

Indeed.

>> So, if you keep the first reaction client running and then start the second. Do you not see the timeout?

The timeout does not occur and both the ReactingClients receive the posted bindingset.

**Comment by Sophietje** [comment]
In GitLab by @barry.nouwt.tno.nl on May 27, 2021, 15:20

Ok, so this is related to the FAQ I posted in this comment: https://gitlab.inesctec.pt/interconnect/knowledge-engine/-/issues/201#note_44932

We should modify the Posting and Reacting clients to also remove their Smart Connectors when shutting down.

Thanks for pointing this out!

As @jarne.kerkaert.inetum-realdolmen.world correctly wondered in issue #214, we are not really maintaining these HTTP clients but used them more as examples. I would recommend creating your own clients based on the REST API OpenAPI spec version 1.7 [here](https://gitlab.inesctec.pt/interconnect/knowledge-engine/-/blob/0.1.7/rest-api/src/main/resources/openapi-sc.yaml).

**Comment by Sophietje** [comment]
In GitLab by @barry.nouwt.tno.nl on May 28, 2021, 06:36

I am closing this issue now. Feel free to reopen it or create a new one if necessary.
